server/src/game/end_game_condition.rs

```diff
@@ -2,26 +2,113 @@ use super::{role_list::Faction, role::Role};
 
 #[derive(Debug, PartialEq, Eq, Clone, Hash)]
 pub enum EndGameCondition {
-    Faction(Faction),
-    Role(Role),
+    Town,
+    Mafia,
+    Cult,
 
-    None
+    Arsonist,
+    Werewolf,
+    Ojo,
+
+    Death,
+    Politician,
+
+    Draw
 }
 impl EndGameCondition {
-    pub fn can_win_together(a: EndGameCondition, b: EndGameCondition)->bool{
-        a == Self::None || b == Self::None || a == b
+    pub fn all()->Vec<EndGameCondition>{
+        vec![
+            EndGameCondition::Town,
+            EndGameCondition::Mafia,
+            EndGameCondition::Cult,
+            EndGameCondition::Arsonist,
+            EndGameCondition::Werewolf,
+            EndGameCondition::Ojo,
+            EndGameCondition::Death,
+            EndGameCondition::Politician
+        ]
+    }
+    ///either return Some(EndGameCondition) or None (if the game is not over yet)
+    pub fn game_is_over(living_roles: Vec<Role>) -> Option<EndGameCondition> {
+
+        //Special wildcard case
+        if living_roles.iter().all(|role|role==&Role::Wildcard)&&living_roles.len()>1{
+            return None;
+        }
+
+        //if nobody is left to hold game hostage
+        if !living_roles.iter().any(|role|EndGameCondition::keeps_game_running(*role)){
+            return Some(EndGameCondition::Draw);
+        }
+
+        //find one end game condition that everyone agrees on
+        for end_game_condition in EndGameCondition::all() {
+            let mut all_agree = true;
+            for living_role in &living_roles {
```

Comment:
I think you should replace this for loop with a `living_roles.iter()...` thing. Personally the way it's written seems a little obtuse -- or at least it was hard for me to understand at first

Reply:
I agree, fix